packages/ledgerstate/branch.go

```diff
@@ -0,0 +1,981 @@
+		err = xerrors.Errorf("unsupported BranchType (%X): %w", branchType, ErrParseBytesFailed)
+		return
+	}
+
+	return
+}
+
+// BranchFromObjectStorage restores a Branch that was stored in the ObjectStorage.
+func BranchFromObjectStorage(_ []byte, data []byte) (branch objectstorage.StorableObject, err error) {
+	if branch, _, err = BranchFromBytes(data); err != nil {
+		err = xerrors.Errorf("failed to parse Branch from bytes: %w", err)
+		return
+	}
+
+	return
+}
+
+// endregion ///////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// region CachedBranch /////////////////////////////////////////////////////////////////////////////////////////////////
+
+// CachedBranch is a wrapper for the generic CachedObject returned by the objectstorage that overrides the accessor
+// methods with a type-casted one.
+type CachedBranch struct {
+	objectstorage.CachedObject
+}
+
+// Retain marks the CachedObject to still be in use by the program.
+func (c *CachedBranch) Retain() *CachedBranch {
+	return &CachedBranch{c.CachedObject.Retain()}
+}
+
+// Unwrap is the type-casted equivalent of Get. It returns nil if the object does not exist.
+func (c *CachedBranch) Unwrap() Branch {
+	untypedObject := c.Get()
+	if untypedObject == nil {
+		return nil
+	}
+
+	typedObject := untypedObject.(Branch)
+	if typedObject == nil || typedObject.IsDeleted() {
+		return nil
+	}
+
+	return typedObject
+}
+
+// Consume unwraps the CachedObject and passes a type-casted version to the consumer (if the object is not empty - it
+// exists). It automatically releases the object when the consumer finishes.
+func (c *CachedBranch) Consume(consumer func(branch Branch), forceRelease ...bool) (consumed bool) {
+	return c.CachedObject.Consume(func(object objectstorage.StorableObject) {
+		consumer(object.(Branch))
+	}, forceRelease...)
+}
+
+// endregion ///////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// region ConflictBranch ///////////////////////////////////////////////////////////////////////////////////////////////
+
+// ConflictBranch represents a container for Transactions and Outputs representing a certain perception of the ledger state.
+type ConflictBranch struct {
+	id             BranchID
+	parents        BranchIDs
+	parentsMutex   sync.RWMutex
+	conflicts      ConflictIDs
+	conflictsMutex sync.RWMutex
+	preferred      bool
+	preferredMutex sync.RWMutex
+	liked          bool
+	likedMutex     sync.RWMutex
+	finalized      bool
+	finalizedMutex sync.RWMutex
+	confirmed      bool
+	confirmedMutex sync.RWMutex
+	rejected       bool
+	rejectedMutex  sync.RWMutex
+
+	objectstorage.StorableObjectFlags
+}
+
+// NewConflictBranch creates a new ConflictBranch from the given details.
+func NewConflictBranch(id BranchID, parents BranchIDs, conflicts ConflictIDs) *ConflictBranch {
+	return &ConflictBranch{
+		id:        id,
+		parents:   parents,
+		conflicts: conflicts,
```

Comment:
Here I think we should copy. Otherwise we need to make sure that `parents` and `conflicts` are not used outside of the just created `ConflictBranch`. This could create some nasty concurrency bugs.